### PR TITLE
codeintel: Add encoding utilities for storing range slices

### PR DESCRIPTION
**Pull request by efritz**
This PR adds utility functions that will later be used when storing denormalized sets of ranges within documents as part of our LSIF -> SCIP pivot.

This encoding uses several observations about the data we'll be storing:

- Each range is represented by a `[start line, start char, end line, end char]` quads
- A slice of ranges are range quads concatenated together
- Each range in the slice is sorted by starting edge
- Reading line numbers from this sequence will form a (mostly, edge cases for overlapping multi-line ranges) non-decreasing sequence
- Character numbers will be bound by some smallish constant (maximum line length in a text document)

The encoding rearranges the set of quads so it's ordered as `[start line 1, end line 1, start line 2, end line 2, ..., start char, end char, ...]`. It then uses delta-encoding to decrease the magnitude of the numbers in the sequence. Then, the delta-encoded sequence is written as a series of varints via Go's binary package.

By scaling out the unit test data, I've found that encoding ranges this way takes only 25% of the storage required by laying out the 32-bit integers. We could probably do better in the future but I think this is a sufficient start (and the 25% figure may be in part due to the pattern in the test data; real-world savings may be better).

## Test plan

New unit tests.

### CODE REVIEW

File: enterprise/internal/codeintel/shared/types/ranges.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+func DecodeRanges(encoded []byte) ([]*scip.Range, error) {
+	flattenedRanges, err := DecodeFlattenedRanges(encoded)
+	if err != nil {
+		return nil, err
+	}
+
+	n := len(flattenedRanges)
+	ranges := make([]*scip.Range, 0, n/4)
+	for i := 0; i < n; i += 4 {
+		ranges = append(ranges, scip.NewRange(flattenedRanges[i:i+4]))
```
Comment (camdencheek) [inline enterprise/internal/codeintel/shared/types/ranges.go:64]: unrelated: it seems non-optimal for `NewRange` to return a pointer for a 16-byte object.

**Review by camdencheek — APPROVED** [review]
Looks great 👍

File: enterprise/internal/codeintel/shared/types/ranges.go (hunk 1 above)
Comment (efritz) [reply enterprise/internal/codeintel/shared/types/ranges.go:69]: I think this repo initially took a very proto-centric encoding. I think we could do better with what's a pointer and what's not (but outside of this PR). Some things will be unmarshaled proto messages, which can't be copied by value, though.

File: enterprise/internal/codeintel/shared/types/ranges.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	// of the loop delta-encodes all of the line numbers as a contiguous sequence. The second
+	// invocation delta-encodes the character offset numbers.
+	//
+	// The delta encoding for line numbers should be impactful as it forms a non-decreasing
+	// sequence, and multiple references to the same variable within a document are likely to
+	// present some degree of locality.
+	//
+	// Delta encoding for character numbers should be impactful as well as the average distance
+	// between character numbers should be less than the average line length in the document. The
+	// vastly common (soft) character maximums fall between 80-120, which fits into a single byte.
```
[the remark this answers is not in the document]
Comment (efritz) [reply enterprise/internal/codeintel/shared/types/ranges.go:44]: [bec22f5](https://github.com/sourcegraph/sourcegraph/pull/44854/commits/bec22f54ef161e35326ffed01205b2cd32fd0e97)

**Pull request merged**